src/compute/src/sink/persist_sink.rs

```diff
@@ -52,92 +58,300 @@ where
+    // workers must write batches with the same boundaries, to ensure that they
+    // can be combined into one batch that gets appended to Consensus state.
     let hashed_id = sink_id.hashed();
-    let active_write_worker = (hashed_id as usize) % scope.peers() == scope.index();
+    let active_worker = (hashed_id as usize) % scope.peers() == scope.index();
+
+    let operator_name = format!("{}-determine_batch_boundaries", operator_name);
+
+    let batch_boundaries =
+        desired_stream.unary_frontier(Pipeline, &operator_name, move |cap, _info| {
+            // Only attempt to write from this frontier onward, as our data are
+            // not necessarily correct for times not greater or equal to this
+            // frontier.
+            let mut current_frontier = as_of;
+            let mut cap_set = CapabilitySet::from_elem(cap);
+
+            move |input, output: &mut OutputHandle<u64, _, Tee<u64, _>>| {
+                while let Some((_time, _data)) = input.next() {
+                    // Just read away data.
+                    // WIP: Is this idiomatic timely?
+                }
+
+                // When the frontier advances, we emit a new description to
+                // signal to the writer operators that they can start writing
+                // updates to persist. In the real-world implementation we might
+                // eventually want to change this to something smarter.
+                if PartialOrder::less_than(&current_frontier.borrow(), &input.frontier().frontier())
+                {
+                    if active_worker {
+                        let write_boundaries = (
+                            current_frontier.clone(),
+                            input.frontier().frontier().to_owned(),
+                        );
+                        let mut session = output.session(cap_set.first().unwrap());
+                        session.give(write_boundaries);
+                    }
+
+                    current_frontier.clear();
+                    current_frontier.extend(input.frontier().frontier().iter().cloned());
+                }
+
+                cap_set.downgrade(input.frontier().frontier().iter());
+            }
+        });
+
+    if sink_id.is_user() {
+        batch_boundaries.inspect(|d| trace!("boundary: {:?}", d));
+    }
+
+    batch_boundaries
+}
+
+/// Writes `desired_stream - persist_stream` to persist, but only for updates
+/// that fall into batch boundaries that we get via `batch_boundaries`. This
+/// forwards a `HollowBatch` for any batch of updates that was written.
+fn write_batches<G>(
+    sink_id: GlobalId,
+    operator_name: String,
+    target: &CollectionMetadata,
+    batch_boundaries: &Stream<G, (Antichain<Timestamp>, Antichain<Timestamp>)>,
+    desired_stream: &Stream<G, (Result<Row, DataflowError>, Timestamp, Diff)>,
+    persist_stream: &Stream<G, (Result<Row, DataflowError>, Timestamp, Diff)>,
```

Comment:
I just realized, multiplexing `Row` with `DataflowError` is pretty bad. A `Row` should have 32 bytes, but a `DataflowError` occupies 96 bytes. Using the multiplexing we inflate the amount of data on the common path by a factor of more than three (32 to 96 bytes, plus the `Result` tag of 8 bytes due to alignment), which is likely to have negative performance implications. If this kind of multiplexing is absolutely necessary, we could consider boxing the error.